concordium-consensus/src/Concordium/Skov/Update.hs

```diff
@@ -512,54 +526,61 @@ doReceiveTransaction tr slot = unlessShutDown $ do
                 statedEnergy = thEnergyAmount $ transactionHeader tx
             if baseEnergy > statedEnergy then return ResultTooLowEnergy
             else do
-               -- this is not ideal since we look up the entire account.
-               -- In the current implementation this is not a problem since all states since the last finalized one are cached
-               -- but this should be revised to add a "accountExists" function in the future
-               senderExists <- flip getAccount (transactionSender tx) =<< queryBlockState =<< lastFinalizedBlock
-               if isNothing senderExists then return ResultNonexistingSenderAccount
-               else snd <$> doReceiveTransactionInternal tr slot
-        _ -> snd <$> doReceiveTransactionInternal tr slot
+               senderExists <- flip accountExists (transactionSender tx) =<< queryBlockState =<< lastFinalizedBlock
+               if not senderExists then return ResultNonexistingSenderAccount
+               else snd <$> doReceiveTransactionInternal True tr (utcTimeToTimestamp now) slot
+        _ -> snd <$> doReceiveTransactionInternal True tr (utcTimeToTimestamp now) slot
     when (ur == ResultSuccess) $ purgeTransactionTable False =<< currentTime
     return ur
-
-    where
-          isExpiryTooLate now = do
-            maxTimeToExpiry <- rpMaxTimeToExpiry <$> getRuntimeParameters
-            let expiry = msgExpiry tr
-            return $ expiry > maxTimeToExpiry + utcTimeToTransactionTime now
+  where
+    isExpiryTooLate tx ts = do
+        maxTimeToExpiry <- rpMaxTimeToExpiry <$> getRuntimeParameters
+        let expiry = msgExpiry tx
+        return $ expiry > maxTimeToExpiry + utcTimeToTransactionTime ts
```

Comment:
If the checks are now more complete and ensure that people can't just spam invalid stuff that would circulate indefinitely we could remove the "too late" check.

Reply:
Yes I agree.